### PR TITLE
feat(Upload body): allow users to upload a dataset body

### PR DESCRIPTION
**Pull request by ramfox**
Users can now drag and drop a file into the body editor! Or they can upload the file by clicking a button.

We plan on eventually getting into a situation where a user can drag over any iteration of a dataset, as a zip or a json, or a section of the dataset (eg meta or transform), and the frontend will unpack it and display it in the editor!

### CODE REVIEW

File: lib/actions/dataset.js [hunk 1]
```diff
@@ -284,17 +284,21 @@ function prepFiles (dataset = {}, transformScript = '', vizScript = '', bodyData
...
-    var name = 'body.'
-    if (dataset && dataset.structure && dataset.structure.format) {
-      name += dataset.structure.format
-    } else {
-      name += 'json'
+  if (bodyData) {
+    if (bodyData.constructor === File) {
+      body = bodyData
+    } else if (bodyData.length) {
+      console.log(bodyData, bodyData.length)
```
Comment (b5) [inline lib/actions/dataset.js:291]: stray console.log

File: lib/components/editor/EditBody.js [hunk 2]
```diff
@@ -82,6 +98,67 @@ export default class EditBody extends Base {
...
+      this.handleSetBodyFile([file])
+      this.setState({ dropZoneState: '' })
+      return
+    }
+    alert('Files must be in csv, json, or cbor format')
+    this.setState({ dropZoneState: '' })
+  }
+
+  handleDragEnd (e) {
+    console.log('drag end')
```
Comment (b5) [inline lib/components/editor/EditBody.js]: stray console.log

File: lib/components/editor/EditBody.js [hunk 3]
```diff
@@ -82,6 +98,67 @@ export default class EditBody extends Base {
     }
   }
 
+  handleDragOver (e) {
+    e.stopPropagation()
+    e.preventDefault()
+    console.log('dragover')
```
Comment (b5) [inline lib/components/editor/EditBody.js]: stray console.log

File: lib/reducers/editor.js [hunk 4]
```diff
@@ -157,13 +157,19 @@ export default function editorReducer (state = initialState, action) {
...
+        // we need to reset the colOrder, rowOrder, and bodyView
         body = undefined
+        colOrder = undefined
+        rowOrder = undefined
+        bodyView = 'json'
       }
-      dataset = {}
-      datasetSections.forEach(s => { dataset[s] = state.dataset[s] })
-      return Object.assign({}, state, { dirty: true, dataset, body })
+      return Object.assign({}, state, { dirty: true, dataset, body, colOrder, rowOrder, bodyView })
```
Comment (b5) [inline lib/reducers/editor.js:172]: great integration with existing body state field!

The only thing I'd also like to see is _clearing_ the body file from the state tree on successful save. We could accomplish this just by sending an `EDITOR_INIT_DATASET` action after a successful save. What matters is we shouldn't hang on to a reference to a file that's potentially large for any longer than we have to. Successful save == we done.

We may later want to consider clearing the file when navigating away, but warning the user before letting them leave the editor, all in the name of preserving how much memory we're consuming for the state tree.

**Review by b5 — APPROVED** [review]
Looks good, one nit

File: lib/reducers/editor.js (hunk 4 above)
Comment (ramfox) [reply lib/reducers/editor.js:172]: dope, good call!

**Pull request merged**